core/src/main/java/jenkins/widgets/WidgetFactory.java

```diff
@@ -0,0 +1,102 @@
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package jenkins.widgets;
+
+import edu.umd.cs.findbugs.annotations.NonNull;
+import hudson.ExtensionList;
+import hudson.ExtensionPoint;
+import hudson.widgets.Widget;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.Set;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import org.kohsuke.accmod.Restricted;
+import org.kohsuke.accmod.restrictions.NoExternalUse;
+
+/**
+ * Allows you to add new widgets to various kind of objects whose class implements {@link HasWidgets}.
+ * @param <T> The type of object to add widgets to.
+ * @param <W> The subtype of {@link Widget} the factory contributes.
+ * @since TODO
+ */
+public abstract class WidgetFactory<T extends HasWidgets, W extends Widget> implements ExtensionPoint {
+    private static final Logger LOGGER = Logger.getLogger(WidgetFactory.class.getName());
+
+    /**
+     * The type of object this factory cares about.
+     * Declared separately, rather than by having {@link #createFor} do a check-cast,
+     * so that method bodies are not loaded until actually needed.
+     * @return the type of {@link T}
+     */
+    public abstract Class<T> type();
+
+    /**
+     * A supertype of any widgets this factory might produce.
+     * Defined so that factories which produce irrelevant widgets need not be consulted.
+     * If your implementation was returning multiple disparate kinds of widgets, it is best to split it into two factories.
+     * <p>If an API defines an abstract {@link Widget} subtype, and you are providing a concrete implementation,
+     * you may return the API type here to delay class loading.
+     * @return a bound for the result of {@link #createFor}
+     */
+    public abstract Class<W> widgetType();
+
+    /**
+     * Creates widgets for a given object.
+     * This may be called frequently for the same object, so if your implementation is expensive, do your own caching.
+     * @param target a widgetable object
+     * @return a possible empty set of widgets (typically either using {@link Set#of}).
+     */
+    public abstract @NonNull Collection<W> createFor(@NonNull T target);
+
+    @Restricted(NoExternalUse.class) // pending a need for it outside HasWidgets
+    public static <T extends HasWidgets, W extends Widget> Iterable<WidgetFactory<T, W>> factoriesFor(Class<T> type, Class<W> widgetType) {
+        List<WidgetFactory<T, W>> result = new ArrayList<>();
+        for (WidgetFactory wf : ExtensionList.lookup(WidgetFactory.class)) {
+            if (wf.type().isAssignableFrom(type) && widgetType.isAssignableFrom(wf.widgetType())) {
+                result.add(wf);
+            }
+        }
+        return result;
+    }
+
+    public Collection<W> createWidgetsFor(HasWidgets hasWidgets) {
```

Comment:
I think this could be restricted too?

Or both could just be inlined into `HasWidgets.getWidgets`.